sympy/tensor/tensor.py

```diff
@@ -2385,6 +2386,32 @@ def _expand_partial_derivative(self):
                     if isinstance(a, TensExpr) else a
                     for a in self.args])
 
+    def _matches_simple(self, expr, repl_dict=None, old=False):
+        """
+        Matches assuming there are no wild objects in self.
+        """
+        if repl_dict is None:
+            repl_dict = {}
+        else:
+            repl_dict = repl_dict.copy()
+
+        if not isinstance(expr, TensExpr):
+            if len(self.get_free_indices()) > 0:
+                #self has indices, but expr does not.
+                return None
+        elif set(self.get_free_indices()) != set(expr.get_free_indices()):
+                #If there are no wilds and the free indices are not the same, they cannot match.
+                return None
+
+        #Since canon_bp may be expensive, we first check for exact equality.
+        diff = self - expr
+        if (diff == 0 or
+            diff.canon_bp() == 0
```

Comment:
Wouldn't this only ever prevent `S.Zero.canon_bp()` from being evaluated? Is that slow in some cases?

The SymPy object for `0` is `S.Zero`

Reply:
I've changed it now since both take under a millisecond.

```
In [8]: %timeit S.Zero == S.Zero
353 ns ± 9.54 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [9]: %timeit canon_bp(S.Zero) == S.Zero
683 ns ± 10.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```